common/web/gesture-recognizer/src/engine/headless/trackedPoint.ts

```diff
@@ -25,68 +27,62 @@ export class TrackedPoint {
-   * @param initialTarget  The initiating event's original target element
+   * @param initialHoveredItem  The initiating event's original target element
    * @param isFromTouch    `true` if sourced from a `TouchEvent`; `false` otherwise.
    */
-  constructor(identifier: number,
-              initialTarget: EventTarget,
-              isFromTouch: boolean);
+  constructor(identifier: number, isFromTouch: boolean) {
+    this.rawIdentifier = identifier;
+    this.isFromTouch = isFromTouch;
+    this._path = new TrackedPath();
+  }
+
   /**
    * Deserializes a TrackedPoint instance from its serialized-JSON form.
+   * @param jsonObj  The JSON representation to deserialize.
    * @param identifier The unique identifier to assign to this instance.
-   * @param parsedObj  The JSON representation to deserialize.
    */
-  constructor(identifier: number,
-    parsedObj: JSONTrackedPoint);
-  constructor(identifier: number,
-              obj: EventTarget | JSONTrackedPoint,
-              isFromTouch?: boolean) {
-    this.rawIdentifier = identifier;
-    if(obj instanceof EventTarget) {
-      this._initialTarget = obj;
-      this.isFromTouch = isFromTouch;
-      this._path = new TrackedPath();
-    } else {
-      // // TEMP:  conversion of old format.
-      // if(obj['sequence']) {
-      //   obj = obj['sequence'];
-      // }
-      // @ts-ignore
-      this.isFromTouch = obj.isFromTouch;
+  public static deserialize(jsonObj: JSONTrackedPoint, identifier: number) {
+    const id = identifier !== undefined ? identifier : this._jsonIdSeed++;
+    const isFromTouch = jsonObj.isFromTouch;
+    const path = TrackedPath.deserialize(jsonObj.path);
+
+    const instance = new TrackedPoint(id, isFromTouch);
+    instance._path = path;
+    return instance;
+  }
 
-      // TEMP:  conversion of old format
-      // @ts-ignore
-      let path = obj.path;
-      // if(obj['samples']) {
-      //   // @ts-ignore
-      //   path = {
-      //     coords: obj['samples']
-      //   };
-      // }
+  public update(sample: InputSample<HoveredItemType>) {
+    this.path.extend(sample);
+  }
 
-      this._path = new TrackedPath(path);
-    }
+  /**
+   * The identifying metadata returned by the configuration's specified `itemIdentifier` for
+   * the target of the first `Event` that corresponded to this `TrackedPoint`.
+   */
+  public get initialHoveredItem(): HoveredItemType {
+    return this.path.coords[0].item;
   }
 
   /**
-   * The event target for the first `Event` corresponding to this `TrackedPoint`.
+   * The identifying metadata returned by the configuration's specified `itemIdentifier` for
+   * the target of the latest `Event` that corresponded to this `TrackedPoint`.
    */
-  public get initialTarget(): EventTarget {
-    return this._initialTarget;
+  public get currentHoveredItem(): HoveredItemType {
+    return this.path.coords[this.path.coords.length-1].item;
```

Comment:
This is going to throw if ever called when the object has an empty path. Is that the failure path we want here, or should it be returning `null` in that case for the caller to deal with?

Reply:
It will never be `null`.

https://github.com/keymanapp/keyman/blob/3fbd59d59dab3111cc96ba3315078b9b898f7c40/common/web/gesture-recognizer/src/engine/headless/trackedPoint.ts#L34-L51

There is no setter; the only way to end up with a nullish value here is if trying to load an improperly-serialized version (when automated testing) that somehow doesn't otherwise throw an error already.

https://github.com/keymanapp/keyman/blob/3fbd59d59dab3111cc96ba3315078b9b898f7c40/common/web/gesture-recognizer/src/engine/headless/trackedPoint.ts#L58-L66

----

Part of me's thinking that maybe I should merge the two classes - `TrackedPoint` and `TrackedPath`.  (Before, only the latter was headless, but now both are.)  I do kind of like having all the path metadata under the `.path` property, though.